includes/class-indieauth-token-endpoint.php

```diff
@@ -192,56 +228,89 @@ public function post( $request ) {
 		return new WP_OAuth_Response( 'invalid_request', __( 'Invalid Request', 'indieauth' ), 400 );
 	}
 
+
+	// Refresh Token Grant Type.
+	public function refresh_token( $params ) {
+
+		$diff = array_diff( array( 'refresh_token' ), array_keys( $params ) );
+		if ( ! empty( $diff ) ) {
+			return new WP_OAuth_Response( 'invalid_request', __( 'The request is missing one or more required parameters', 'indieauth' ), 400 );
+		}
+		$refresh = $this->refresh_tokens->get( $params['refresh_token'] );
+		if ( ! $refresh ) {
+			return new WP_OAuth_Response( 'invalid_grant', __( 'Invalid Token', 'indieauth' ), 400 );
+		}
+
+		// Destroy the refresh token.
+		$this->refresh_tokens->destroy( $params['refresh_token'] );
+
+		if ( array_key_exists( 'client_id', $params ) && ( $refresh['client_id'] !== $params['client_id'] ) ) {
+			return new WP_OAuth_Response( 'invalid_grant', __( 'Client ID does not match', 'indieauth' ), 400 );
+		}
+
+		return $this->generate_token_response( $refresh );
+
+	}
+
 	// Authorization Code Grant Type.
 	public function authorization_code( $params ) {
 		$diff = array_diff( array( 'code', 'client_id', 'redirect_uri' ), array_keys( $params ) );
 		if ( ! empty( $diff ) ) {
 			return new WP_OAuth_Response( 'invalid_request', __( 'The request is missing one or more required parameters', 'indieauth' ), 400 );
 		}
-		$args     = array_filter(
+		$args                  = array_filter(
 			array(
 				'code'          => $params['code'],
 				'redirect_uri'  => $params['redirect_uri'],
 				'client_id'     => $params['client_id'],
 				'code_verifier' => isset( $params['code_verifier'] ) ? $params['code_verifier'] : null,
 			)
 		);
-		$response = indieauth_verify_local_authorization_code( $args );
+		$response              = indieauth_verify_local_authorization_code( $args );
+		$response['client_id'] = $params['client_id'];
```

Comment:
is this not stored as part of the authorization code?

Reply:
It is.